crates/bevy_ecs/src/bundle.rs

```diff
@@ -781,27 +801,29 @@ impl<'w> BundleInserter<'w> {
             }
         };
 
+        let new_archetype = &*new_archetype;
         // SAFETY: We have no outstanding mutable references to world as they were dropped
         let mut deferred_world = unsafe { self.world.into_deferred() };
 
-        if new_archetype.has_on_add() {
-            // SAFETY: All components in the bundle are guaranteed to exist in the World
-            // as they must be initialized before creating the BundleInfo.
-            unsafe {
-                deferred_world.trigger_on_add(
-                    entity,
-                    bundle_info
-                        .iter_components()
-                        .zip(add_bundle.bundle_status.iter())
-                        .filter(|(_, &status)| status == ComponentStatus::Added)
-                        .map(|(id, _)| id),
+        // SAFETY: All components in the bundle are guaranteed to exist in the World
+        // as they must be initialized before creating the BundleInfo.
+        unsafe {
+            deferred_world.trigger_on_add(new_archetype, entity, add_bundle.added.iter().cloned());
+            if new_archetype.has_add_observer() {
+                deferred_world.trigger_observers(
+                    ON_ADD,
+                    Some(entity),
+                    add_bundle.added.iter().cloned(),
+                );
+            }
```

Comment:
The difference for observers is that they don't have specific functions for each of add, insert and remove since they also need to cater to other in-built events and user events, I could create wrapper functions that call trigger internally though.

Reply:
I see, no strong opinion then 🙂 
If you think this represents the difference better let's keep it